flink-core/src/main/java/org/apache/flink/migration/util/MigrationInstantiationUtil.java

```diff
@@ -47,9 +47,16 @@ public ClassLoaderObjectInputStream(InputStream in, ClassLoader classLoader) thr
 
 			// the flink package may be at position 0 (regular class) or position 2 (array)
 			final int flinkPackagePos;
-			if ((flinkPackagePos = className.indexOf(FLINK_BASE_PACKAGE)) == 0 ||
-					(flinkPackagePos == 2 && className.startsWith(ARRAY_PREFIX)))
-			{
+			if (className.contains("org.apache.flink.runtime.state.ArrayListSerializer")) {
```

Comment:
The code here is a little tricky. I think we should use a Map to record all modified classes and their corresponding backups.